tempesta_fw/http.c

```diff
@@ -563,24 +615,28 @@ tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp)
 		goto send_500;
 
 	/* Add request to the server connection. */
-	spin_lock(&conn->msg_qlock);
-	list_add_tail(&req->msg.msg_list, &conn->msg_queue);
-	spin_unlock(&conn->msg_qlock);
+	spin_lock(&srv_conn->msg_qlock);
+	list_add_tail(&req->msg.msg_list, &srv_conn->msg_queue);
+	spin_unlock(&srv_conn->msg_qlock);
 
 	/* Send request to the server. */
-	tfw_connection_send(conn, (TfwMsg *)req, false);
-
+	tfw_connection_send(srv_conn, (TfwMsg *)req, false);
 	goto conn_put;
+
 send_404:
 	tfw_http_send_404((TfwHttpMsg *)req);
+	if (req_conn_close)
+		tfw_connection_drop(cli_conn);
 	tfw_http_conn_msg_free((TfwHttpMsg *)req);
```

Comment:
The 3 lines above are ugly (see https://github.com/natsys/tempesta/pull/361/files#r47531757). You introduce ```tfw_connection_drop()``` just to call ```tfw_sock_clnt_drop()``` (for client, the server logic is the same). Meantime, ```tfw_http_conn_msg_free()``` calls ```tfw_cli_conn_release()``` after reference counter decrement. So we're in the same code place (sock_clnt.c) but, through 2 functions and using callbacks through SS. The design is too complicated and potentially buggy.